### PR TITLE
James 3945 rights positioning for subaddressing

### PR DESCRIPTION
**Pull request by florentos17**
i am working on implementing sub-addressing for apache james.
for now, i am trying to make right positioning work with IMAP.
the final goal is to make it work for JMAP !
this is still work in progress.

any comment is welcome :)

### CODE REVIEW

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/SubAddressingTest.java [hunk 1]
```diff
@@ -97,43 +97,106 @@ void tearDown() throws IOException {
...
-    void subAddressedEmailShouldBeDeliveredInINBOXWhenNoRights(@TempDir File temporaryFolder) throws Exception {
+    void subAddressedEmailShouldBeDeliveredInINBOXWhenNobodyHasRight(@TempDir File temporaryFolder) throws Exception {
         setup(temporaryFolder);
 
         // create mailbox
-        testIMAPClient.connect(LOCALHOST_IP, jamesServer.getProbe(ImapGuiceProbe.class).getImapPort())
-            .login(RECIPIENT, PASSWORD)
-            .create(TARGETED_MAILBOX);
+        testIMAPClient.sendCommand("CREATE " + TARGETED_MAILBOX);
+        assertThat(testIMAPClient.sendCommand("GETACL " + TARGETED_MAILBOX)).contains("OK GETACL");
```
Comment (chibenwa) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/SubAddressingTest.java:113]: Tests are supposed to get only a single assertion. THis allow making clear what is tested.

If intermediate states are unsure, this likely means a distinct test shall be added, covering explicitly the effects of SETACL. Also those very valuable tests have not much to do with subadressing and shall be done in dedicated location (RIghts.script in mpt-core ?)

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/SubAddressingTest.java [hunk 2]
```diff
@@ -97,43 +97,106 @@ void tearDown() throws IOException {
...
 
-        //do not give posting rights
+        // do not give posting rights
+        testIMAPClient.sendCommand("SETACL " + TARGETED_MAILBOX + " " + "anybody" + " -p");
```
Comment (chibenwa) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/SubAddressingTest.java:116]: The RFC says `anyone` and not `anybody`. We shall be spec compliant! Likely the IMAP parsing layers shall be revised.

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/SubAddressingTest.java (hunk 1 above)
Comment (chibenwa) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/SubAddressingTest.java:113]: (You are not asserting anything as `OK GETACL` will always be returned. An assertion proving the ACL is empty would be asserting that the only line in the IMAP response is `OK GETACL`.

**Review by chibenwa — COMMENTED** [review]
This goes into the right direction, but detailed tests onto the ACL system deserves to be located in mpt imap-mailbox Rights.script and not mixed up with subaddressing feature. IMAP parsing likely need to be reworked not to confuse anybody with anyone...

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/SubAddressingTest.java (hunk 1 above)
Comment (florentos17) [reply server/mailet/integration-testing/src/test/java/org/apache/james/mailets/SubAddressingTest.java:113]: yes you're right about the extra assertions, they are just for me "to be sure" but they are technically useless.

about the dedicated location, I can't find Rights.script or mpt-core ; would mpt/impl/imap-mailbox/inmemory/src/test/java/org/apache/james/mpt/imapmailbox/inmemory/InMemoryPositionRightTest.java be ok ?

Comment (chibenwa) [reply server/mailet/integration-testing/src/test/java/org/apache/james/mailets/SubAddressingTest.java:113]: For the third time I share the link of the existing test file for Right management on Apache James 

https://github.com/apache/james-project/blob/master/mpt/impl/imap-mailbox/core/src/main/resources/org/apache/james/imap/scripts/Rights.test

File: mpt/impl/imap-mailbox/core/src/main/resources/org/apache/james/imap/scripts/Rights.test [hunk 3]
```diff
@@ -38,3 +38,12 @@ S: a6 NO DELETEACL failed. Mailbox not found.
 
 C: a7 SETACL INBOX bob lr
 S: a7 OK SETACL completed.
+
+C: a8 SETACL INBOX anyone p
+S: a8 OK SETACL completed.
```
Comment (chibenwa) [inline mpt/impl/imap-mailbox/core/src/main/resources/org/apache/james/imap/scripts/Rights.test:43]: You are not getting the ACL after each SETACL in order to asser the changed done?

File: mpt/impl/imap-mailbox/core/src/main/resources/org/apache/james/imap/scripts/Rights.test [hunk 4]
```diff
@@ -38,3 +38,12 @@ S: a6 NO DELETEACL failed. Mailbox not found.
...
+
+C: a8 SETACL INBOX anyone -p
+S: a8 OK SETACL completed.
+
+C: a8 SETACL INBOX -bob p
+S: a8 OK SETACL completed.
```
Comment (chibenwa) [inline mpt/impl/imap-mailbox/core/src/main/resources/org/apache/james/imap/scripts/Rights.test:49]: What if I `SETACL bob p` and then after `SETACL -bob p` what is the final state of the ACL? is bob allowed?

Idem with the reverse.

File: mpt/impl/imap-mailbox/inmemory/src/test/java/org/apache/james/mpt/imapmailbox/inmemory/InMemoryPositionRightTest.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+ *   "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *   KIND, either express or implied.  See the License for the
+ *   specific language governing permissions and limitations
+ *   under the License.
+ *
+ */
+
+package org.apache.james.mpt.imapmailbox.inmemory;
+
+public class InMemoryPositionRightTest {
```
Comment (chibenwa) [inline mpt/impl/imap-mailbox/inmemory/src/test/java/org/apache/james/mpt/imapmailbox/inmemory/InMemoryPositionRightTest.java:23]: Useless?

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/SubAddressingTest.java [hunk 6]
```diff
@@ -113,90 +115,21 @@ void subAddressedEmailShouldBeDeliveredInINBOXWhenNobodyHasRight(@TempDir File t
...
-        // whitelist sender 2 and send from sender 1
-        testIMAPClient.sendCommand("SETACL " + TARGETED_MAILBOX + " " + FROM2 + " +p");
-        assertThat(testIMAPClient.sendCommand("GETACL " + TARGETED_MAILBOX)).contains("\"user4@james.org\" \"p\"");
-
-        sendSubAddressedMail();
-        awaitSubAddressedMail(MailboxConstants.INBOX);
-    }
-
-    @Test
-    void subAddressedEmailShouldBeDeliveredInINBOXWhenSenderIsBlacklisted(@TempDir File temporaryFolder) throws Exception {
```
Comment (chibenwa) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/SubAddressingTest.java]: Why removing all those tests?

Comment (florentos17) [reply server/mailet/integration-testing/src/test/java/org/apache/james/mailets/SubAddressingTest.java]: you said that they had not much to do with subaddressing and shall be done in a dedicated location, so I thought those weren't needed anymore ?

**Review by chibenwa — APPROVED** [review]
Looks fairly good to me!

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/SubAddressingTest.java (hunk 6 above)
Comment (chibenwa) [reply server/mailet/integration-testing/src/test/java/org/apache/james/mailets/SubAddressingTest.java]: No, the intermediate assertions have nothing to do with SubAdressing but the overall test looks ok.

**Comment by chibenwa** [comment]
Still a draft?

**Comment by florentos17** [comment]
> Still a draft?

i am waiting to see if the tests finally pass after this morning's fixups...!

**Comment by Arsnael** [comment]
https://ci-builds.apache.org/job/james/job/ApacheJames/job/PR-2419/7/#showFailuresLink

Related.

**Review by Arsnael — APPROVED** [review]
Overall looks ok to me though. just need to fix the failing tests

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
Common practice on the project and with git:

Fixups should start with `fixup! [COMMIT]`.

Like if you have a commit `JAMES-3945 rights positioning with IMAP in integration tests` a fixup for this commit would be:
`fixup! JAMES-3945 rights positioning with IMAP in integration tests`

Makes it easier for reviewers to know what is being fixed and also easier to squash up and have have a clean git history merged.

It's ok I think to squash all those commits into one here so I will merge, just think about it for the next time :)

**Comment by florentos17** [comment]
> Common practice on the project and with git:
> 
> Fixups should start with `fixup! [COMMIT]`.
> 
> Like if you have a commit `JAMES-3945 rights positioning with IMAP in integration tests` a fixup for this commit would be: `fixup! JAMES-3945 rights positioning with IMAP in integration tests`
> 
> Makes it easier for reviewers to know what is being fixed and also easier to squash up and have have a clean git history merged.
> 
> It's ok I think to squash all those commits into one here so I will merge, just think about it for the next time :)

Yes, I intended to squash the fixups once the tests would pass and before merging, but you're right anyway, thank you !

**Pull request merged**